### ISSUE TITLE
Use Push notifications

### ISSUE THREAD

**Issue by davidak** [nucleus-ffm/foss_warn#27]
As it can be critical that notifications are delivered instantly, Push notifications should be used when technically possible. (Android with Google Services or microG)

If not possible, tell the user that it takes up to 15 minutes to manually check and offer instructions to install microG (it is very hard, maybe not viable).

This might decide between live and death of a user or their family!

**Comment by nucleus-ffm** [comment]
It would be really cool to have that, but for any kind of push services (maybe with a websocket) I would need a webserver, but I don't have one. And even if I were to rent one, I'm not sure I'd have the time and skills to maintain the server all the time. Also, I had to think about how to finance the server.

**Comment by davidak** [comment]
I understand it's not easy. Maybe a solution can be found.

Maybe you can find contributors who love to maintain a server.

Would funding help? The project might have a good chance to get funding from [Prototype Fund](https://prototypefund.de/about/public-interest-tech/).

**Comment by jugendhacker** [comment]
Maybe [unified push](https://unifiedpush.org/developers/flutter/) could be used so that it will still be compatible with F-Droid? Fluffychat does this for example.

**Comment by budda85** [comment]
There is für example [Ntfy](https://ntfy.sh/) .
It is very easy to install. Fluffychat has automatic create an pushservice with ntfy.

**Comment by jugendhacker** [comment]
@budda85 yeah Fluffy just uses unified push, which Ntfy happens to be one provider for...

**Comment by citizenserious** [comment]
[NextPush](https://unifiedpush.org/users/distributors/nextpush/)  would be an option for me and maybe for other too. It is implemented in [fedilab](https://fedilab.app/)  and [Element](https://github.com/vector-im/element-android/)  as an example (:

**Comment by bt90** [comment]
I'm not sure if an ntfy server alone would cut it. Aren't the UnifiedPush endpoints autogenerated? e.g. https://ntfy.sh/upMg6xDhy2UGUW

So we would need an additional service to handle registration and publishing for these endpoints.

Being able to register a UP receiver for a predefined server and topic would avoid that.

@binwiedehier is this possible?

**Comment by karmanyaahm** [comment]
Yes, just UnifiedPush alone won't do. Currently, the app talks to the government API every 15 minutes. But, for push, there would need to be a server that talks to that API, perhaps every 15 seconds, and then pushes to everyone registered on that server.

@vkrause Is currently experimenting with a server like that for KDE, and since it sends data over UnifiedPush, it should work even for an Android app.

See: https://floss.social/@VolkerKrause/109477562343881737
And the server: https://invent.kde.org/vkrause/kpublicalerts/

**Comment by ChaosNicro** [comment]
I have seen people successfully parse server notifications to an https endpoint, tricking the server into sending to a UP distributor. But with no insight into NINAs communication protocol, that wont be viable.

**Comment by nucleus-ffm** [comment]
@ChaosNicro I am not sure I understood your post correctly. Can you give me an example of someone who has done this?

**Comment by nucleus-ffm** [comment]
Btw, here's a quick update on this topic. 

We are currently experimenting with a push server written with NodeJS that uses unifiedPush to manage subscriptions and send push notifications. The server is of course also open source https://github.com/nucleus-ffm/foss_warn_server. At the moment it is still very simple and limited to the bare essentials - just for experimenting. The Flutter part is also in progress in the https://github.com/nucleus-ffm/foss_warn/tree/push-service branch. Since the whole server part is new to me, it would be great if someone would help me with it 

However, there is a big problem at the moment. The Flutter connector https://github.com/UnifiedPush/flutter-connector is not compatible with the AlarmManager at the moment. When the AlarmManager is initialized, the callback functions of the Flutter connector for e.g. new Message are not called. I have already created an issue for this https://github.com/UnifiedPush/flutter-connector/issues/123 If someone here has an idea how to fix this, maybe they can help to solve the problem.

**Comment by ChaosNicro** [comment]
> @ChaosNicro I am not sure I understood your post correctly. Can you give me an example of someone who has done this?

Sorry, I was drawing from https://github.com/Telegram-FOSS-Team/Telegram-FOSS/issues/577#issuecomment-1285792906
Like I said, probably not an option. The idea is to impersonate a (NINA)-Client and have it send updates to a given UP-Server when registering the client. But that's not easy with proprietary apps. It's just odd that a public service like this doesn't just provide an open push-option, but just a polling-API.

**Comment by p1gp1g** [comment]
@nucleus-ffm I have tested with `unifiedpush:^5.0.0` and it seems the issue is resolved. Could you try it ?

**Comment by nucleus-ffm** [comment]
@p1gp1g I tried it and as before, as soon as I initialise the AlarmManager, the callbacks stop working. I tried `unifiedpush:^5.0.0` and `android_alarm_manager_plus: ^2.1.4` with the Unified Example App. @MatsG23 maybe you could try this too?

**Comment by p1gp1g** [comment]
I have probably been too fast with my tests with Foss_Warn. Sorry for the ping

**Comment by nucleus-ffm** [comment]
@MatsG23 Oh yes, sorry. I forgot to check if the db folder existed and if not, to create one. I have corrected this and pushed the changes. Also the FOSS Warn part is now updated. It should work now.
You will need to install a distributor. I use [ntfy]( https://f-droid.org/de/packages/io.heckel.ntfy/)

How to use the server:
- Start the server (I use nodemon for this). 
- Start FOSS Warn (and for now edit lines 47 and 48 in main.dart to disable the alarmManager).
- Select a new location in FOSS Warn
- The data is sent to the server
- The server should now output the client table with the new entry.
- You can now comment out line 96 in server.js to run `jsonWorker.loadJson()` and restart the server.
- The server should now call the NINA API (only once for now) and send a push notification if needed.

The Python scripts are only for testing the server. They just send a dummy request to the server to check if it works as expected. If you want to test the server, just run the restRegistration.py file and you should get a "success" back.

**Comment by PureTryOut** [comment]
So the app uses UnifiedPush for notifications nowadays through the use of FOSS Public Alert Server (FPAS), replacing the custom FOSSWarn specific server. With that, I'm closing this issue.
